container_checker of monit sanity check failed

<!--
If you are reporting a new issue, make sure that we do not have any duplicates
already open. You can ensure this by searching the issue list for this
repository. If there is a duplicate, please close your issue and add a comment
to the existing issue instead.

If you suspect your issue is a bug, please edit your issue description to
include the BUG REPORT INFORMATION shown below. If you fail to provide this
information within 7 days, we cannot debug your issue and will close it. We
will, however, reopen it if you later provide the information.

For more information about reporting issues, see
https://github.com/Azure/SONiC/wiki#report-issues

---------------------------------------------------
GENERAL SUPPORT INFORMATION
---------------------------------------------------

The GitHub issue tracker is for bug reports and feature requests.
General support can be found at the following locations:

- SONiC Support Forums - https://groups.google.com/forum/#!forum/sonicproject

---------------------------------------------------
BUG REPORT INFORMATION
---------------------------------------------------
Use the commands below to provide key information from your environment:
You do NOT have to include this information if this is a FEATURE REQUEST
-->

**Description**

<!--
Briefly describe the problem you are having in a few paragraphs.
-->

**Steps to reproduce the issue:**
1. Run a test with sanity check enabled.
2. Monit check failed with "container_checker": "Status failed",
3.

**Describe the results you received:**

```
_____________________________________________________________________________ ERROR at setup of test_bgp_gr_helper_routes_perserved _____________________________________________________________________________
    
localhost = <tests.common.devices.local.Localhost object at 0x7f84650164d0>, duthosts = <tests.common.devices.duthosts.DutHosts object at 0x7f8464c734d0>
request = <SubRequest 'sanity_check' for <Function test_bgp_gr_helper_routes_perserved>>
fanouthosts = {u'str-7060cx-32s-21': { os: 'eos', hostname: 'str-7060cx-32s-21', device_type: 'FanoutLeaf' }, u'str-7060cx-32s-22': { os: 'eos', hostname: 'str-7060cx-32s-22', device_type: 'FanoutLeaf' }}
tbinfo = {'auto_recover': 'True', 'comment': 'nejo', 'conf-name': 'vms3-t1-dx010-1', 'duts': ['str-dx010-acs-1'], ...}
                                                       
    @pytest.fixture(scope="module", autouse=True)
    def sanity_check(localhost, duthosts, request, fanouthosts, tbinfo):       
        logger.info("Prepare sanity check")
    
        skip_sanity = False         
        allow_recover = False                                                                        
        recover_method = "adaptive"                                       
        pre_check_items = set(copy.deepcopy(SUPPORTED_CHECKS))  # Default check items                          
        post_check = False
                                                                                                              
        customized_sanity_check = None
        for m in request.node.iter_markers():
            logger.info("Found marker: m.name=%s, m.args=%s, m.kwargs=%s" % (m.name, m.args, m.kwargs))
            if m.name == "sanity_check":                                                                                                                    
                customized_sanity_check = m                                                             
                break
                         
        if customized_sanity_check:                         
            logger.info("Process marker {} in script. m.args={}, m.kwargs={}"
                .format(customized_sanity_check.name, customized_sanity_check.args, customized_sanity_check.kwargs))
            skip_sanity = customized_sanity_check.kwargs.get("skip_sanity", False)                                      
            allow_recover = customized_sanity_check.kwargs.get("allow_recover", False)
            recover_method = customized_sanity_check.kwargs.get("recover_method", "adaptive")
            if allow_recover and recover_method not in constants.RECOVER_METHODS:
                pytest.warning("Unsupported recover method")
                logger.info("Fall back to use default recover method 'config_reload'")
                recover_method = "config_reload"                
                                             
            pre_check_items = _update_check_items(                                                                                                                                                                               pre_check_items,               
                customized_sanity_check.kwargs.get("check_items", []),
                SUPPORTED_CHECKS)
                                                                                     
            post_check = customized_sanity_check.kwargs.get("post_check", False)                    
                                                                                                                                                                                                                         if request.config.option.skip_sanity:                                        
            skip_sanity = True                                                                                                                                                                              
        if skip_sanity:              
            logger.info("Skip sanity check according to command line argument or configuration of test script.")
            yield                                                     
            return                                                                        
                          
        if request.config.option.allow_recover:                                                  
            allow_recover = True                   
                                                                                           
        if request.config.option.post_check:                                       
            post_check = True                   
                                                                                                                          
        cli_check_items = request.config.getoption("--check_items")
        cli_post_check_items = request.config.getoption("--post_check_items")
                                                                                                                                                                                                                 
        if cli_check_items:                                                                                                                  
            logger.info('Fine tune pre-test check items based on CLI option --check_items')                                                                                                                      
            cli_items_list=str(cli_check_items).split(',')    
            pre_check_items = _update_check_items(pre_check_items, cli_items_list, SUPPORTED_CHECKS)
    
        pre_check_items = filter_check_items(tbinfo, pre_check_items)  # Filter out un-supported checks.
    
        if post_check:
            # Prepare post test check items based on the collected pre test check items.
            post_check_items = copy.copy(pre_check_items)
            if customized_sanity_check:
                post_check_items = _update_check_items(
                    post_check_items,
                    customized_sanity_check.kwargs.get("post_check_items", []),
                    SUPPORTED_CHECKS)
    
            if cli_post_check_items:
                logger.info('Fine tune post-test check items based on CLI option --post_check_items')
                cli_post_items_list = str(cli_post_check_items).split(',')
                post_check_items = _update_check_items(post_check_items, cli_post_items_list, SUPPORTED_CHECKS)

            post_check_items = filter_check_items(tbinfo, post_check_items)  # Filter out un-supported checks.
        else:                                                 
            post_check_items = set()                            
                                                           
        logger.info("Sanity check settings: skip_sanity=%s, pre_check_items=%s, allow_recover=%s, recover_method=%s, post_check=%s, post_check_items=%s" % \
            (skip_sanity, pre_check_items, allow_recover, recover_method, post_check, post_check_items))
                                                         
        for item in pre_check_items.union(post_check_items):  
            request.fixturenames.append(_item2fixture(item))
                                                            
            # Workaround for pytest requirement.         
            # Each possibly used check fixture must be executed in setup phase. Otherwise there could be teardown error.
            request.getfixturevalue(_item2fixture(item))      
                                                         
        if pre_check_items:  
            logger.info("Start pre-test sanity checks")
                     
            # Dynamically attach selected check fixtures to node
            for item in set(pre_check_items):
                request.fixturenames.append(_item2fixture(item))                                                                                                                                                 
                                               
            print_logs(duthosts)
                            
            check_results = do_checks(request, pre_check_items, stage=STAGE_PRE_TEST)
            logger.debug("Pre-test sanity check results:\n%s" % json.dumps(check_results, indent=4))
                                                                                                                                                                                                                 
            failed_results = [result for result in check_results if result['failed']]
            if failed_results:                                                                                                                                                                              
                if not allow_recover:
                    pt_assert(False, "!!!!!!!!!!!!!!!!Pre-test sanity check failed: !!!!!!!!!!!!!!!!\n{}"\
>                       .format(json.dumps(failed_results, indent=4)))
E                   Failed: !!!!!!!!!!!!!!!!Pre-test sanity check failed: !!!!!!!!!!!!!!!!
E                   [      
E                       {                                                                        
E                           "check_item": "monit", 
E                           "failed": true,                                                
E                           "host": "str-dx010-acs-1",                             
E                           "services_status": {
E                               "acms|dSMS_endpoint_modifier": "Status ok",                                               
E                               "swss|intfmgrd": "Status ok",      
E                               "swss|portmgrd": "Status ok",                
E                               "lldp|lldpmgrd": "Status ok",                                                                                                                                                    
E                               "acms|cert_converter": "Status ok",                                                                          
E                               "lldp|lldpd_monitor": "Status ok",                                                                                                                                               
E                               "swss|coppmgrd": "Status ok",                                                                                                                                                   
E                               "swss|portsyncd": "Status ok", 
E                               "swss|buffermgrd": "Status ok", 
E                               "snmp|snmp_subagent": "Status ok", 
E                               "database|redis_server": "Status ok", 
E                               "swss|vxlanmgrd": "Status ok", 
E                               "teamd|teamsyncd": "Status ok", 
E                               "telemetry|dialout_client": "Status ok", 
E                               "swss|neighsyncd": "Status ok", 
E                               "rsyslog": "Running", 
E                               "root-overlay": "Accessible", 
E                               "swss|nbrmgrd": "Status ok", 
E                               "bgp|fpmsyncd": "Status ok", 
E                               "str-dx010-acs-1": "Running", 
E                               "telemetry|telemetry": "Status ok", 
E                               "bgp|bgpcfgd": "Status ok", 
E                               "swss|orchagent": "Status ok", 
E                               "snmp|snmpd": "Status ok", 
E                               "swss|vrfmgrd": "Status ok", 
E                               "teamd|teammgrd": "Status ok", 
E                               "syncd|dsserve": "Status ok", 
E                               "swss|tunnelmgrd": "Status ok", 
E                               "bgp|bgpmon": "Status ok", 
E                               "lldp|lldp_syncd": "Status ok", 
E                               "container_checker": "Status failed", 
E                               "var-log": "Accessible", 
E                               "swss|vlanmgrd": "Status ok", 
E                               "bgp|staticd": "Status ok", 
E                               "syncd|syncd": "Status ok", 
E                               "bgp|bgpd": "Status ok", 
E                               "bgp|zebra": "Status ok", 
E                               "swss|fdbsyncd": "Status ok", 
E                               "routeCheck": "Status ok"
E                           }
E                       }
E                   ]
```

The output of command `sudo monit status` contains below information:
```
Program 'container_checker'
  status                       Status failed
  monitoring status            Monitored
  monitoring mode              active
  on reboot                    start
  last exit value              3
  last output                  Expected containers not running: mux
  data collected               Thu, 01 Apr 2021 06:51:09
```

The DUT is running t1-lag topology. It is not a dualtor setup.

```
admin@str-dx010-acs-1:~$ docker ps -a
CONTAINER ID        IMAGE                             COMMAND                  CREATED             STATUS              PORTS               NAMES
ea4aa389f4c3        docker-sonic-telemetry:latest     "/usr/local/bin/supe鈥   About an hour ago   Up 17 minutes                           telemetry
e69ac63615c0        docker-snmp:latest                "/usr/local/bin/supe鈥   About an hour ago   Up 16 minutes                           snmp
42efa4f23e2e        docker-router-advertiser:latest   "/usr/bin/docker-ini鈥   About an hour ago   Up 16 minutes                           radv
b4ed0f70cfe0        docker-dhcp-relay:latest          "/usr/bin/docker_ini鈥   About an hour ago   Up 17 minutes                           dhcp_relay
3419b7b27b7e        docker-lldp:latest                "/usr/bin/docker-lld鈥   About an hour ago   Up 17 minutes                           lldp
17acf7002d92        docker-syncd-brcm:latest          "/usr/local/bin/supe鈥   About an hour ago   Up 17 minutes                           syncd
aaca6d68d64a        docker-teamd:latest               "/usr/local/bin/supe鈥   About an hour ago   Up 17 minutes                           teamd
0c9e217bcbec        docker-platform-monitor:latest    "/usr/bin/docker_ini鈥   About an hour ago   Up 17 minutes                           pmon
626df73ecf13        docker-orchagent:latest           "/usr/bin/docker-ini鈥   About an hour ago   Up 17 minutes                           swss
d5a8863608ee        docker-fpm-frr:latest             "/usr/bin/docker_ini鈥   About an hour ago   Up 17 minutes                           bgp
c3d40e5f1417        docker-acms:latest                "/usr/local/bin/supe鈥   About an hour ago   Up About an hour                        acms
6bb94e8fbc2a        docker-database:latest            "/usr/local/bin/dock鈥   About an hour ago   Up About an hour                        database
```

**Describe the results you expected:**
Sanity check should pass.

**Additional information you deem important:**
<!--
software/ASIC/Hardware Flatform version and info
-->
    **Output of `show version`:**

```
admin@str-dx010-acs-1:~$ show ver

SONiC Software Version: SONiC.20201231.beta.16
Distribution: Debian 10.9
Kernel: 4.19.0-12-2-amd64
Build commit: 1423f7a3e
Build date: Tue Mar 30 08:58:29 UTC 2021
Built by: sonicbld@w1

Platform: x86_64-cel_seastone-r0
HwSKU: Celestica-DX010-C32
ASIC: broadcom
ASIC Count: 1
Serial Number: DX010F2B018B03BY100009
Uptime: 06:50:04 up 47 min,  1 user,  load average: 3.10, 4.42, 4.76

Docker images:
REPOSITORY                 TAG                 IMAGE ID            SIZE
docker-syncd-brcm          20201231.beta.16    129cd4fad177        695MB
docker-syncd-brcm          latest              129cd4fad177        695MB
docker-teamd               20201231.beta.16    28a46d30f94a        414MB
docker-teamd               latest              28a46d30f94a        414MB
docker-router-advertiser   20201231.beta.16    c19ddea99334        404MB
docker-router-advertiser   latest              c19ddea99334        404MB
docker-platform-monitor    20201231.beta.16    ad643bfdb1ec        612MB
docker-platform-monitor    latest              ad643bfdb1ec        612MB
docker-snmp                20201231.beta.16    471d7f50a763        444MB
docker-snmp                latest              471d7f50a763        444MB
docker-database            20201231.beta.16    317c8f871ffb        404MB
docker-database            latest              317c8f871ffb        404MB
docker-lldp                20201231.beta.16    400faf3db5fe        444MB
docker-lldp                latest              400faf3db5fe        444MB
docker-orchagent           20201231.beta.16    286a65357f1c        432MB
docker-orchagent           latest              286a65357f1c        432MB
docker-dhcp-relay          20201231.beta.16    585004d83e1c        411MB
docker-dhcp-relay          latest              585004d83e1c        411MB
docker-sonic-telemetry     20201231.beta.16    52cd156926b3        493MB
docker-sonic-telemetry     latest              52cd156926b3        493MB
docker-mux                 20201231.beta.16    19f75d3ad40a        458MB
docker-mux                 latest              19f75d3ad40a        458MB
docker-fpm-frr             20201231.beta.16    b3fa7e41628b        432MB
docker-fpm-frr             latest              b3fa7e41628b        432MB
docker-acms                20201231.beta.16    eb77a3f5aa6a        197MB
docker-acms                latest              eb77a3f5aa6a        197MB
k8s.gcr.io/pause           3.2                 80d28bedfe5d        683kB
```

    **Attach debug file `sudo generate_dump`:**

    ```
    (paste your output here)
    ```

--- a/files/image_config/monit/container_checker
+++ b/files/image_config/monit/container_checker
@@ -48,11 +48,11 @@ def get_command_result(command):
 def get_expected_running_containers():
     """
     @summary: This function will get the expected running containers by following the rule:
-              The 'state' field of container in 'FEATURE' table should not be 'disabled'. Then
-              if the device has Multi-ASIC, this function will get container list by determining the
+              The 'state' field of container in 'FEATURE' table should not be 'disabled' or 'always_disabled'.
+              If the device has Multi-ASIC, this function will get container list by determining the
               value of field 'has_global_scope', the number of ASICs and the value of field
-              'has_per_asic_scope'. If the device has single ASIC, the container name was put into
-              the list.
+              'has_per_asic_scope'.
+              If the device has single ASIC, the container name was put into the list.
     @return:  A set which contains the expected running containers.
     """
     config_db = swsssdk.ConfigDBConnector()
@@ -62,7 +62,8 @@ def get_expected_running_containers():
     expected_running_containers = set()
 
     for container_name in feature_table.keys():
-        if feature_table[container_name]["state"] != "disabled":
+        if (feature_table[container_name]["state"] != "disabled" and
+                feature_table[container_name]["state"] != "always_disabled"):
             if multi_asic.is_multi_asic():
                 if feature_table[container_name]["has_global_scope"] == "True":
                     expected_running_containers.add(container_name)

--- a/files/image_config/monit/container_checker
+++ b/files/image_config/monit/container_checker
@@ -62,8 +62,7 @@ def get_expected_running_containers():
     expected_running_containers = set()
 
     for container_name in feature_table.keys():
-        if (feature_table[container_name]["state"] != "disabled" and
-                feature_table[container_name]["state"] != "always_disabled"):
+        if feature_table[container_name]["state"] not in ["disabled", "always_disabled"]:
             if multi_asic.is_multi_asic():
                 if feature_table[container_name]["has_global_scope"] == "True":
                     expected_running_containers.add(container_name)

